### PR TITLE
Added a counter for low-level inputs (e.g. TPs within TPSets) to TriggerGenericMaker

### PR DESCRIPTION
**Pull request by bieryAtFnal**
… so that it can be reported at Stop time.

I found this debug information helpful when trying to understand why the number of TAMaker outputs occasionally fluctuates down.

### CODE REVIEW

File: src/trigger/TriggerGenericMaker.hpp [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include "appfwk/DAQModule.hpp"
 #include "appfwk/DAQModuleHelper.hpp"
 #include "daqdataformats/SourceID.hpp"
+#include "dfmessages/Types.hpp"
```
Comment (MRiganSUSX) [inline src/trigger/TriggerGenericMaker.hpp:21]: If I understand correctly, you want this for the run number. Would it be better to use daqdataformats/Types.hpp instead, as that is what  dfmessages/Types.hpp refers to anyway?

File: src/trigger/TriggerGenericMaker.hpp [hunk 2]
```diff
@@ -113,6 +115,7 @@ class TriggerGenericMaker : public dunedaq::appfwk::DAQModule
   using metric_counter_type = decltype(triggergenericmakerinfo::Info::received_count);
   std::atomic<metric_counter_type> m_received_count;
   std::atomic<metric_counter_type> m_sent_count;
+  dfmessages::run_number_t m_run_number;
```
Comment (MRiganSUSX) [inline src/trigger/TriggerGenericMaker.hpp:118]: same as above

File: src/trigger/TriggerGenericMaker.hpp [hunk 3]
```diff
@@ -231,17 +236,20 @@ class TriggerGenericWorker
 {
 public:
   explicit TriggerGenericWorker(TriggerGenericMaker<IN, OUT, MAKER>& parent)
-    : m_parent(parent)
+    : m_parent(parent), m_low_level_input_count(0)
```
Comment (MRiganSUSX) [inline src/trigger/TriggerGenericMaker.hpp:239]: I think this template (single object IN, single object OUT) is unused in the trigger (just a comment, your expansion should definitely be included)

File: src/trigger/TriggerGenericMaker.hpp [hunk 4]
```diff
@@ -340,6 +354,7 @@ class TriggerGenericWorker<Set<A>, Set<B>, MAKER>
             // This should never happen, but we check here so we at least get some output if it did
             ers::fatal(OutOfOrderSets(ERS_HERE, m_parent.get_name(), end_time, in.start_time));
           }
+          m_low_level_input_count += time_slice.size();
```
Comment (MRiganSUSX) [inline src/trigger/TriggerGenericMaker.hpp:357]: idea: any point in not mixing payloads and heatbeats?

**Review by MRiganSUSX — COMMENTED** [review]
Hi Kurt,
Thanks for this addition. I think it is very useful. 
I have nothing against this implementation, just the comments below and 2 more questions:
-> could this be made part of opmon? (or, would that be helpful) 
-> would there be any advantage of having this part of the parent class (m_parent, similar as for m_sent_count)? This would open things like aggregating multiple AMs together ?

**Comment by MRiganSUSX** [comment]
Other comment:
I have tested this with a simple setup (1 APA, 1 maker) and more complicated setup (3 APAs, 2 makers), both worked and produced (reasonable) counts:

```
tam_1_0: Exiting do_work() method for run 1, received 12916 inputs (69808 sub-inputs) and successfully sent 2700 outputs. 
tam_1_1: Exiting do_work() method for run 1, received 13323 inputs (94160 sub-inputs) and successfully sent 618 outputs. 
tam_2_0: Exiting do_work() method for run 1, received 12495 inputs (81836 sub-inputs) and successfully sent 3100 outputs. 
tam_2_1: Exiting do_work() method for run 1, received 13234 inputs (93622 sub-inputs) and successfully sent 614 outputs. 
tam_3_0: Exiting do_work() method for run 1, received 12453 inputs (84840 sub-inputs) and successfully sent 3200 outputs. 
tam_3_1: Exiting do_work() method for run 1, received 13191 inputs (93204 sub-inputs) and successfully sent 612 outputs. 
tcm_0: Exiting do_work() method for run 1, received 6543 inputs (6543 sub-inputs) and successfully sent 262 outputs. 
tcm_1: Exiting do_work() method for run 1, received 1844 inputs (1844 sub-inputs) and successfully sent 1843 outputs. 
```

I'm happy with this being merged. There are some comments on possible extensions.

**Comment by bieryAtFnal** [comment]
Thanks Michal for the suggestions.
As you suggest, we'll go ahead and merge these changes in now, and make notes about the suggestions for the future.

**Pull request merged**